core/src/ping_pong.rs

```diff
@@ -0,0 +1,401 @@
+    pub fn add(&mut self, pong: &Pong, socket: SocketAddr, now: Instant) -> bool {
+        let node = (pong.pubkey(), socket);
+        match self.pending_cache.peek(&pong.hash) {
+            Some(value) if *value == node => {
+                self.pings.pop(&node);
+                self.pongs.put(node, now);
+                self.pending_cache.pop(&pong.hash);
+                true
+            }
+            _ => false,
+        }
+    }
+
+    /// Checks if the remote node has been pinged recently. If not, calls the
+    /// given function to generates a new ping message, records current
+    /// timestamp and hash of ping token, and returns the ping message.
+    fn maybe_ping<T, F>(
+        &mut self,
+        now: Instant,
+        node: (Pubkey, SocketAddr),
+        mut pingf: F,
+    ) -> Option<Ping<T>>
+    where
+        T: Serialize,
+        F: FnMut() -> Option<Ping<T>>,
+    {
+        // Rate limit consecutive pings sent to a remote node.
+        let delay = self.ttl / 64;
+        match self.pings.peek(&node) {
+            Some(t) if now.saturating_duration_since(*t) < delay => None,
+            _ => {
+                let ping = pingf()?;
+                let hash = hash::hash(&serialize(&ping.token).ok()?);
+                self.pings.put(node, now);
+                self.pending_cache.put(hash, node);
+                Some(ping)
+            }
+        }
+    }
+
+    /// Returns true if the remote node has responded to a ping message.
+    /// Removes expired pong messages. In order to extend verifications before
+    /// expiration, if the pong message is not too recent, and the node has not
+    /// been pinged recently, calls the given function to generates a new ping
+    /// message, records current timestamp and hash of ping token, and returns
+    /// the ping message.
+    /// Caller should verify if the socket address is valid. (e.g. by using
+    /// ContactInfo::is_valid_address).
+    pub fn check<T, F>(
+        &mut self,
+        now: Instant,
+        node: (Pubkey, SocketAddr),
+        pingf: F,
+    ) -> (bool, Option<Ping<T>>)
+    where
+        T: Serialize,
+        F: FnMut() -> Option<Ping<T>>,
+    {
+        let (check, should_ping) = match self.pongs.get(&node) {
+            None => (false, true),
+            Some(t) => {
+                let age = now.saturating_duration_since(*t);
+                // Pop if the pong message has expired.
+                if age > self.ttl {
+                    self.pongs.pop(&node);
+                }
+                // If the pong message is not too recent, generate a new ping
+                // message to extend remote node verification.
+                (true, age > self.ttl / 8)
+            }
+        };
+        let ping = if should_ping {
+            self.maybe_ping(now, node, pingf)
+        } else {
+            None
+        };
+        (check, ping)
+    }
+}
+
+impl Clone for PingCache {
+    fn clone(&self) -> Self {
```

Comment:
Is it just for tests?

Reply:
This is needed for `ClusterInfo::clone_with_id` function:
https://github.com/solana-labs/solana/blob/e666272ab/core/src/cluster_info.rs#L477-L499
which is only for tests and simulations.

I have to implement this manually because `LruCache` does not implement `Clone`.

Reply:
I removed the `Clone` implementation in favor of a separate `mock_clone` function which is more explicit and emphasizes that it is only for tests and simulations. Also, updated the clone code because it has to traverse the iterators in reverse order to maintain the original order of insertions.